From 39aa9cbd1eea8cb37eb3c04f50034d17431b0870 Mon Sep 17 00:00:00 2001
From: "chak-kei.lam" <chak-kei.lam@mediatek.com>
Date: Fri, 6 Jun 2025 22:24:10 +0800
Subject: [PATCH] net: ethernet: mtk_ppe: add roaming handler

Without this patch, the traffic may stop over 30 seconds when the
WiFi roams.
---
 drivers/net/ethernet/mediatek/mtk_eth_soc.c   |   7 +
 drivers/net/ethernet/mediatek/mtk_eth_soc.h   |   5 +
 drivers/net/ethernet/mediatek/mtk_ppe.c       | 211 ++++++++++++++++++
 drivers/net/ethernet/mediatek/mtk_ppe.h       |   2 +
 .../net/ethernet/mediatek/mtk_ppe_debugfs.c   |  56 +++++
 5 files changed, 281 insertions(+)

diff --git a/drivers/net/ethernet/mediatek/mtk_eth_soc.c b/drivers/net/ethernet/mediatek/mtk_eth_soc.c
index 35ef427..92b0c6c 100644
--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.c
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.c
@@ -4144,6 +4144,11 @@ static int mtk_open(struct net_device *dev)
 			mtk_ppe_start(eth->ppe[i]);
 #endif
 
+		err = mtk_ppe_roaming_start(eth);
+		if (err)
+			netdev_err(dev, "%s: could not start ppe roaming work: %d\n",
+				   __func__, err);
+
 		for (i = 0; i < MTK_MAX_DEVS; i++) {
 			if (!eth->netdev[i])
 				continue;
@@ -4299,6 +4304,8 @@ static int mtk_stop(struct net_device *dev)
 	if (MTK_HAS_CAPS(eth->soc->caps, MTK_HWTSTAMP))
 		mtk_ptp_clock_deinit(eth);
 
+	mtk_ppe_roaming_stop(eth);
+
 #if !defined(CONFIG_NET_MEDIATEK_HNAT) && !defined(CONFIG_NET_MEDIATEK_HNAT_MODULE)
 	for (i = 0; i < ARRAY_SIZE(eth->ppe); i++)
 		mtk_ppe_stop(eth->ppe[i]);
diff --git a/drivers/net/ethernet/mediatek/mtk_eth_soc.h b/drivers/net/ethernet/mediatek/mtk_eth_soc.h
index 9aff6ab..25bfada 100644
--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.h
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.h
@@ -1684,10 +1684,15 @@ struct mtk_eth {
 
 	struct metadata_dst		*dsa_meta[MTK_MAX_DSA_PORTS];
 
+	u8				debug_level;
 	u8				qos_toggle;
 	struct mtk_ppe			*ppe[3];
 	struct rhashtable		flow_table;
 
+	struct socket			*ppe_roam_sock;
+	struct work_struct		ppe_roam_work;
+	unsigned char			ppe_roam_buf[1024];
+
 	struct bpf_prog			__rcu *prog;
 
 	struct {
diff --git a/drivers/net/ethernet/mediatek/mtk_ppe.c b/drivers/net/ethernet/mediatek/mtk_ppe.c
index 193093b..1b198cb 100644
--- a/drivers/net/ethernet/mediatek/mtk_ppe.c
+++ b/drivers/net/ethernet/mediatek/mtk_ppe.c
@@ -10,6 +10,7 @@
 #include <linux/if_vlan.h>
 #include <net/dst_metadata.h>
 #include <net/dsa.h>
+#include <net/netfilter/nf_flow_table.h>
 #include "mtk_eth_soc.h"
 #include "mtk_ppe.h"
 #include "mtk_ppe_regs.h"
@@ -525,6 +526,28 @@ mtk_flow_entry_match(struct mtk_eth *eth, struct mtk_flow_entry *entry,
 	return !memcmp(&entry->data.data, &data->data, len - 4);
 }
 
+static bool mtk_foe_mac_match(struct mtk_foe_entry *entry, u8 *mac)
+{
+	int type;
+
+	type = FIELD_GET(MTK_FOE_IB1_PACKET_TYPE, entry->ib1);
+	if (type > MTK_PPE_PKT_TYPE_IPV4_DSLITE) {
+		if(((swab32(entry->ipv6.l2.dest_mac_hi) == *(u32 *)mac) &&
+		    (swab16(entry->ipv6.l2.dest_mac_lo) == *(u16 *)&mac[4])) ||
+		   ((swab32(entry->ipv6.l2.src_mac_hi) == *(u32 *)mac) &&
+		    (swab16(entry->ipv6.l2.src_mac_lo) == *(u16 *)&mac[4])))
+			return true;
+	} else {
+		if(((swab32(entry->ipv4.l2.dest_mac_hi) == *(u32 *)mac) &&
+		    (swab16(entry->ipv4.l2.dest_mac_lo) == *(u16 *)&mac[4])) ||
+		   ((swab32(entry->ipv4.l2.src_mac_hi) == *(u32 *)mac) &&
+		    (swab16(entry->ipv4.l2.src_mac_lo) == *(u16 *)&mac[4])))
+			return true;
+	}
+
+	return false;
+}
+
 static void
 __mtk_foe_entry_clear(struct mtk_ppe *ppe, struct mtk_flow_entry *entry,
 		      bool set_state)
@@ -1132,3 +1155,191 @@ int mtk_ppe_stop(struct mtk_ppe *ppe)
 
 	return 0;
 }
+
+static int
+mtk_ppe_check_wdma_path(struct mtk_eth *eth, struct mtk_foe_entry *foe)
+{
+	u32 sp, winfo;
+	u32 *ib2;
+
+	ib2 = mtk_foe_entry_ib2(eth, foe);
+	sp = mtk_get_ib1_sp(eth, foe);
+	winfo = FIELD_GET(MTK_FOE_IB2_WDMA_WINFO, *ib2);
+
+	if (winfo || (sp == PSE_WDMA0_PORT ||
+		      sp == PSE_WDMA1_PORT ||
+		      sp == PSE_WDMA2_PORT))
+		return 1;
+
+	return 0;
+}
+
+static int
+mtk_flow_offload_teardown_by_roaming(struct mtk_ppe *ppe, u8 *mac)
+{
+	int i, j, count = 0;
+
+	if (!ppe)
+		return 0;
+
+	for (i = 0; i < MTK_PPE_ENTRIES; i++) {
+		struct mtk_foe_entry *entry = mtk_foe_get_entry(ppe, i);
+		struct flow_offload_tuple tuple;
+		int type, state;
+
+		state = FIELD_GET(MTK_FOE_IB1_STATE, entry->ib1);
+		if (state != MTK_FOE_STATE_BIND || !mtk_foe_mac_match(entry, mac) ||
+		    !mtk_ppe_check_wdma_path(ppe->eth, entry))
+			continue;
+
+		memset(&tuple, 0, sizeof(tuple));
+
+		if (entry->ib1 & MTK_FOE_IB1_UDP)
+			tuple.l4proto = IPPROTO_UDP;
+		else
+			tuple.l4proto = IPPROTO_TCP;
+
+		type = FIELD_GET(MTK_FOE_IB1_PACKET_TYPE, entry->ib1);
+		if (type > MTK_PPE_PKT_TYPE_IPV4_DSLITE) {
+			tuple.l3proto = NFPROTO_IPV6;
+			tuple.src_port = htons(entry->ipv6.src_port);
+			tuple.dst_port = htons(entry->ipv6.src_port);
+			for (j = 0; j < 4; j ++) {
+				tuple.src_v6.s6_addr32[j] = htonl(entry->ipv6.src_ip[j]);
+				tuple.dst_v6.s6_addr32[j] = htonl(entry->ipv6.dest_ip[j]);
+			}
+		} else {
+			tuple.l3proto = NFPROTO_IPV4;
+			tuple.src_port = htons(entry->ipv4.orig.src_port);
+			tuple.dst_port = htons(entry->ipv4.orig.dest_port);
+			tuple.src_v4.s_addr = htonl(entry->ipv4.orig.src_ip);
+			tuple.dst_v4.s_addr = htonl(entry->ipv4.orig.dest_ip);
+		}
+
+		flow_offload_teardown_by_tuple(&tuple);
+		count++;
+
+		pr_info("mtk_ppe: the roamiing entry (%x) has been deleted\n", i);
+	}
+
+	if (!count && ppe->eth->debug_level >= 4)
+		pr_warn("mtk_ppe: the roaming entry was not found in the PPE(%d)\n",
+			ppe->index);
+
+	return count;
+}
+
+static void mtk_ppe_roam_handler(struct work_struct *work)
+{
+	struct mtk_eth *eth = container_of(work, struct mtk_eth, ppe_roam_work);
+	struct net_device *dev = NULL;
+	struct kvec iov;
+	struct msghdr msg;
+	struct nlmsghdr *nlh;
+	struct ndmsg *ndm;
+	struct nlattr *nla;
+	int len, ifindex, i;
+	u8 mac[ETH_ALEN];
+
+	if (!eth->ppe_roam_sock)
+		return;
+
+	iov.iov_base = eth->ppe_roam_buf;
+	iov.iov_len = sizeof(eth->ppe_roam_buf);
+	memset(&msg, 0, sizeof(msg));
+	msg.msg_namelen = sizeof(struct sockaddr_nl);
+
+	len = kernel_recvmsg(eth->ppe_roam_sock, &msg, &iov, 1, iov.iov_len, 0);
+	if (len <= 0)
+		goto out;
+
+	nlh = (struct nlmsghdr*)eth->ppe_roam_buf;
+	if (!NLMSG_OK(nlh, len) || nlh->nlmsg_type != RTM_NEWNEIGH)
+		goto out;
+
+	len = nlh->nlmsg_len - NLMSG_HDRLEN;
+	ndm = (struct ndmsg *)NLMSG_DATA(nlh);
+	if (ndm->ndm_family != PF_BRIDGE)
+		goto out;
+
+	ifindex = ndm->ndm_ifindex;
+	dev = dev_get_by_index(&init_net, ifindex);
+	if (!dev)
+		goto out;
+
+	if (!dev->ieee80211_ptr)
+		goto out;
+
+	nla = (struct nlattr *)((unsigned char *)ndm + sizeof(struct ndmsg));
+	len -= NLMSG_LENGTH(sizeof(struct ndmsg));
+	while (nla_ok(nla, len)) {
+		if (nla_type(nla) == NDA_LLADDR) {
+			memcpy(mac, nla_data(nla), ETH_ALEN);
+			for (i = 0; i < ARRAY_SIZE(eth->ppe); i++)
+				mtk_flow_offload_teardown_by_roaming(eth->ppe[i], mac);
+			if (eth->debug_level >= 6)
+				pr_info("mtk_ppe: the neighbor (%pM) has been updated\n", mac);
+		}
+		nla = nla_next(nla, &len);
+	}
+
+out:
+	if (dev)
+		dev_put(dev);
+	if (!work_pending(&eth->ppe_roam_work))
+		schedule_work(&eth->ppe_roam_work);
+}
+
+int mtk_ppe_roaming_start(struct mtk_eth *eth)
+{
+	struct socket *sock = NULL;
+	struct sockaddr_nl addr;
+	int ret;
+
+	INIT_WORK(&eth->ppe_roam_work, mtk_ppe_roam_handler);
+
+	ret = sock_create_kern(&init_net, AF_NETLINK, SOCK_RAW, NETLINK_ROUTE, &sock);
+	if (ret < 0) {
+		pr_warn("mtk_ppe: roaming work is unable to create socket\n");
+		goto out;
+	}
+
+	addr.nl_family = AF_NETLINK;
+	addr.nl_pad = 0;
+	addr.nl_pid = 65534;
+	addr.nl_groups = 1 << (RTNLGRP_NEIGH - 1);
+	ret = kernel_bind(sock, (struct sockaddr *)&addr, sizeof(addr));
+	if (ret < 0) {
+		pr_warn("mtk_ppe: roaming work is unable to bind socket\n");
+		goto out;
+	}
+
+	eth->ppe_roam_sock = sock;
+	eth->ppe_roam_sock->sk->sk_rcvtimeo = msecs_to_jiffies(100);
+
+	schedule_work(&eth->ppe_roam_work);
+
+	pr_info("mtk_ppe: roaming work has been activated\n");
+
+	return 0;
+
+out:
+	if (sock)
+		sock_release(sock);
+
+	return ret;
+}
+
+int mtk_ppe_roaming_stop(struct mtk_eth *eth)
+{
+	if (!eth->ppe_roam_sock)
+		return -ENOENT;
+
+	cancel_work_sync(&eth->ppe_roam_work);
+	sock_release(eth->ppe_roam_sock);
+	eth->ppe_roam_sock = NULL;
+
+	pr_info("mtk_ppe: roaming work has been deactivated\n");
+
+	return 0;
+}
diff --git a/drivers/net/ethernet/mediatek/mtk_ppe.h b/drivers/net/ethernet/mediatek/mtk_ppe.h
index 87b8223..e06d757 100644
--- a/drivers/net/ethernet/mediatek/mtk_ppe.h
+++ b/drivers/net/ethernet/mediatek/mtk_ppe.h
@@ -350,6 +350,8 @@ struct mtk_ppe {
 struct mtk_ppe *mtk_ppe_init(struct mtk_eth *eth, void __iomem *base, int index);
 
 void mtk_ppe_deinit(struct mtk_eth *eth);
+int mtk_ppe_roaming_start(struct mtk_eth *eth);
+int mtk_ppe_roaming_stop(struct mtk_eth *eth);
 void mtk_ppe_start(struct mtk_ppe *ppe);
 int mtk_ppe_stop(struct mtk_ppe *ppe);
 int mtk_ppe_prepare_reset(struct mtk_ppe *ppe);
diff --git a/drivers/net/ethernet/mediatek/mtk_ppe_debugfs.c b/drivers/net/ethernet/mediatek/mtk_ppe_debugfs.c
index 129bde4..9a494a3 100644
--- a/drivers/net/ethernet/mediatek/mtk_ppe_debugfs.c
+++ b/drivers/net/ethernet/mediatek/mtk_ppe_debugfs.c
@@ -251,6 +251,60 @@ static const struct file_operations mtk_ppe_internal_debugfs_qos_fops = {
 	.release = single_release,
 };
 
+static int
+mtk_ppe_internal_debugfs_level_show(struct seq_file *m, void *private)
+{
+	struct mtk_eth *eth = m->private;
+
+	seq_printf(m, "PPE debug level=%d\n", eth->debug_level);
+
+	return 0;
+}
+
+static int
+mtk_ppe_internal_debugfs_level_open(struct inode *inode, struct file *file)
+{
+	return single_open(file, mtk_ppe_internal_debugfs_level_show,
+			   inode->i_private);
+}
+
+static ssize_t
+mtk_ppe_internal_debugfs_level_write(struct file *file, const char __user *buf,
+				     size_t count, loff_t *offset)
+{
+	struct seq_file *m = file->private_data;
+	struct mtk_eth *eth = (struct mtk_eth *)m->private;
+	char tmp[8] = {0};
+	u32 level;
+
+	if ((count > sizeof(tmp)))
+		return -EFAULT;
+
+	if (copy_from_user(tmp, buf, count))
+		return -EFAULT;
+
+	if (sscanf(tmp, "%d", &level) != 1)
+		return -EFAULT;
+
+	if (level < 0 || level > 7) {
+		pr_warn("The input debug level is invalid, "
+			"it should range from 0 to 7.\n");
+		return -EINVAL;
+	}
+
+	eth->debug_level = level;
+
+	return count;
+}
+
+static const struct file_operations mtk_ppe_internal_debugfs_level_fops = {
+	.open = mtk_ppe_internal_debugfs_level_open,
+	.read = seq_read,
+	.llseek = seq_lseek,
+	.write = mtk_ppe_internal_debugfs_level_write,
+	.release = single_release,
+};
+
 static int
 mtk_ppe_internal_debugfs_foe_all_show(struct seq_file *m, void *private)
 {
@@ -310,6 +364,8 @@ int mtk_ppe_internal_debugfs_init(struct mtk_eth *eth)
 			    &mtk_ppe_internal_debugfs_foe_bind_fops);
 	debugfs_create_file("qos_toggle", S_IRUGO, root, eth,
 			    &mtk_ppe_internal_debugfs_qos_fops);
+	debugfs_create_file("debug_level", S_IRUGO, root, eth,
+			    &mtk_ppe_internal_debugfs_level_fops);
 
 	for (i = 0; i < (mtk_is_netsys_v2_or_greater(eth) ? 4 : 2); i++) {
 		ret = snprintf(name, sizeof(name), "qdma_sch%ld", i);
-- 
2.45.2

